### ISSUE TITLE
Exact Order of Crafting Steps 

### ISSUE THREAD

**Issue by 9yte** [wronnyhuang/metapoison#1]
Hi,

Thanks for sharing the source code for this amazing project. This helps a lot.
I apologize if it's not the right place to ask this question, but I thought that might be somebody else's question in the future.

To me, the precise order of operations in each crafting step is not clear. Not that your code is not clear, I have almost zero experience with TensorFlow and that's why I'm having a hard time understanding some parts of the code. It would be great if, for the following example, you explain what is the exact sequence of operations taken in one step of poison crafting. 

Let's assume we have 120 samples in our training sample, the batch size is 40. We have 20 poisons, 8 in the first batch, 12 in the second batch, and none in the last batch. The number of SGD unroll steps is 2 (K=2). We have 3 substitute models (M=3).

My understanding is that these are the exact order of operations in each step of poison crafting:

```
meta_loss = 0
for model in subs_models:
    T = model.parameters

    T0 = clone(T)
    # First round of unrolling for all (i.e., three) batches of the training set
    T1 = T0 - a * grad_T(loss(X1, Y1, T0))
    T2 = T1 - a * grad_T(loss(X2, Y2, T1))
    T3 = T2 - a * grad_T(loss(X3, Y3, T2))

    # Second round of unrolling
    T4 = T3 - a * grad_T(loss(X1, Y1, T3))
    T5 = T4 - a * grad_T(loss(X2, Y2, T4))
    T6 = T5 - a * grad_T(loss(X3, Y3, T5))

    # Now T6 is a function of poisons
    meta_loss += adv_loss(x_t, y_adv, T6)

    # Train the original model for one more epoch, or if the maximum epoch reached, reset the model.
    model = train_or_reset(model, X, Y)

XP_1 = XP_0 - b * grad_XP(meta_loss)
```
 
Where X1, X2, and X3 show three batches for the training inputs, and Y1, Y2, and Y3 show the corresponding training labels. XP represents all the 20 poison samples.

I'm not sure if my interpretation of Algorithm 1 is correct or not. If not, it would be great if you provide me with the correct steps for this example.

Thanks a lot for your help.

**Comment by wronnyhuang** [comment]
Hi @9yte,

Sorry for the late reply and thanks for the careful question. The algorithm has some differences from the psuedocode you've presented. Let me show you the correct version using the notation and settings you've proposed.

```
meta_grad = 0
for model in subs_models:

    # if the maximum epoch reached, reset the model.
    model = reset_if_exceeded(model)

    T = model.parameters

    T0 = clone(T)
    # First round of unrolling (K=2, so we unroll only twice)
    T1 = T0 - a * grad_T(loss(X1, Y1, T0))
    T2 = T1 - a * grad_T(loss(X2', Y2, T1)) # X2' = stop_grad(X2)
    meta_grad += grad_XP(adv_loss(x_t, y_adv, T2)) # grads for 8 out of 20 poisons will be nonzero since loss only depends on X1 which contains 8 poisons
    update(T) # train SGD step on current batch. This is where the model training actually happens. Doing so between each unrolling round helps diversify the models that the poisons experience

    T0 = clone(T)
    # Second round of unrolling
    T1 = T0 - a * grad_T(loss(X2, Y2, T0))
    T2 = T1 - a * grad_T(loss(X3', Y3, T1)) # X3' = stop_grad(X3)
    meta_grad += grad_XP(adv_loss(x_t, y_adv, T2)) # grads for 12 out of 20 poisons will be nonzero since loss only depends on X2 which contains 12 poisons
    update(T)

    T0 = clone(T)
    # Third round of unrolling
    # Unrolling would be skipped since X3 happens to contain no poisons, but otherwise it'd look like this
    T1 = T0 - a * grad_T(loss(X3, Y3, T0))
    T2 = T1 - a * grad_T(loss(X1', Y1, T1)) # X1' = stop_grad(X1)
    meta_grad += grad_XP(adv_loss(x_t, y_adv, T2)) # grads for 0 out of 20 poisons will be nonzero since loss only depends on X3 which contains 0 poisons
    update(T)

XP_1 = XP_0 - b * meta_grad / num_subs_models
```

So you see that if K=2, then there's only 2 grad_T's computed. Also, we accumulate the metagradients rather than the metalosses.

**Comment by 9yte** [comment]
Thanks a lot for your response. Oh, many things now make sense! I just have one question left: Do you use the same ordering of samples across all models (in one particular step)? According to the above example, it seems yes.

**Comment by wronnyhuang** [comment]
The order of samples in the minibatches is random for each model
